include/unifex/v2/async_scope.hpp

```diff
@@ -191,55 +191,102 @@ struct _nest_op final {
 template <typename Sender, typename Receiver>
 using nest_op = typename _nest_op<Sender, Receiver>::type;
 
+template <typename Sender, typename Receiver>
+struct _nest_receiver final {
+  struct type;
+};
+
+template <typename Sender, typename Receiver>
+using nest_receiver = typename _nest_receiver<Sender, Receiver>::type;
+
 template <typename Sender, typename Receiver>
 struct _nest_op<Sender, Receiver>::type final {
   template <typename Sender2, typename Receiver2>
   explicit type(Sender2&& s, Receiver2&& r, scope_reference&& scope) noexcept(
-      is_nothrow_connectable_v<Sender2, Receiver2>)
-    : scope_(std::move(scope)) {
+      is_nothrow_connectable_v<Sender2, nest_receiver<Sender, Receiver>>)
+    : scope_(std::move(scope))
+    , receiver_(static_cast<Receiver2&&>(r)) {
     UNIFEX_ASSERT(scope_);
     activate_union_member_with(op_, [&]() {
       return unifex::connect(
-          static_cast<Sender2&&>(s), static_cast<Receiver2&&>(r));
+          static_cast<Sender2&&>(s), nest_receiver<Sender, Receiver>{this});
     });
   }
 
   explicit type(Receiver&& r) noexcept(
-      std::is_nothrow_move_constructible_v<Receiver>) {
-    activate_union_member(receiver_, std::move(r));
-  }
+      std::is_nothrow_move_constructible_v<Receiver>)
+    : receiver_(std::move(r)) {}
 
   explicit type(const Receiver& r) noexcept(
-      std::is_nothrow_copy_constructible_v<Receiver>) {
-    activate_union_member(receiver_, r);
-  }
+      std::is_nothrow_copy_constructible_v<Receiver>)
+    : receiver_(r) {}
 
   type(type&& op) = delete;
 
-  ~type() {
-    if (scope_) {
-      deactivate_union_member(op_);
-    } else {
-      deactivate_union_member(receiver_);
-    }
-  }
+  ~type() { cleanup(); }
 
   friend void tag_invoke(tag_t<start>, type& op) noexcept {
     if (op.scope_) {
       unifex::start(op.op_.get());
     } else {
-      unifex::set_done(std::move(op).receiver_.get());
+      unifex::set_done(std::move(op).receiver_);
     }
   }
 
-private:
-  using op_t = connect_result_t<Sender, Receiver>;
+  void cleanup() noexcept {
+    if (auto scope = std::move(scope_)) {
+      op_.destruct();
+    }
+  }
+
+  using op_t = connect_result_t<Sender, nest_receiver<Sender, Receiver>>;
 
   scope_reference scope_;
-  union {
-    UNIFEX_NO_UNIQUE_ADDRESS manual_lifetime<Receiver> receiver_;
-    UNIFEX_NO_UNIQUE_ADDRESS manual_lifetime<op_t> op_;
-  };
+  UNIFEX_NO_UNIQUE_ADDRESS Receiver receiver_;
+  UNIFEX_NO_UNIQUE_ADDRESS manual_lifetime<op_t> op_;
+};
+
+template <typename Sender, typename Receiver>
+struct _nest_receiver<Sender, Receiver>::type final {
+  nest_op<Sender, Receiver>* op_;
+
+  template <typename... T>
+  void set_value(T... values) noexcept {
+    auto op = op_;
+
+    op->cleanup();
```

Comment:
`op->cleanup()` destroys the nested operation state _and drops the scope reference_.

Perhaps the scope reference should be dropped later?  If one of `values...` is a pointer to a thing protected by the scope, the pointer could dangle by the time `op->receiver_` observes it.  Maybe `cleanup()` should return-by-move the operation's `scope_ref` so that we can drop the reference after  `set_value()` returns?

I might be over thinking this.

Reply:
After chatting with Ján, we've decided I should defer dropping the reference until after we're done invoking completion methods.